Fix geojson identifier

Conengmo opened the pull request:
For styling and highlights we need a unique field on each geojson feature. This functionality was broken for geojson data with certain properties layouts. This change makes the functionality work for all properties allowed by the geojson spec.

Add a utility function to do a dynamic key lookup in a dictionary.

Add test cases for all changes.

Closes #1154.

Conengmo commented:
This part of the code is becoming a bit extensive, I'm doubtful if it's worth it.

We could instead decide not to support using properties as feature identifiers. That way, a feature either needs to have an id or data should be embedded.

Alternatively there are two other changes possible:
- don't require a feature identifier if the style function applies to all features equally.
- translate the style function to a Javascript function, so no other feature identifier is needed.

ocefpaf commented:
> This part of the code is becoming a bit extensive, I'm doubtful if it's worth it.

Probably not but I'll leave the call to you b/c you will be here longer than me :smile: 

> We could instead decide not to support using properties as feature identifiers. That way, a feature either needs to have an id or data should be embedded so we can add an id.

I'm OK with that.

> Alternatively there are two other changes possible:
> 
> * don't require a feature identifier if the style function applies to all features equally.
> * translate the style function to a Javascript function, so no other feature identifier is needed.

I'm not sure I like either of these but again, it is you call.

Conengmo commented:
I simplified it to the following cases:

- Each feature has a unique id.
- Each feature has a properties dict with a unique string or int value.

I think these two cases are well defined, so they're easy to test. I added multiple tests, mainly to verify we recover from getting faulty data.
